data-processing/common/upload_entities.py

```diff
@@ -155,73 +109,193 @@ def upload_entities(
     # should happen after the entity models are created, because the relationships may include
     # references between entities (e.g., some symbols may be children of others), and we need to
     # know the row IDs of the uploaded entities to make links between them in the database.
-    uploaded_entity_models: Dict[Tuple[str, str], Entity] = {}
+    uploaded_entities = fetch_existing_entities(s2_id, data_version)
+
+    entity_relationship_models: List[EntityDataModel] = []
+    for entity in entities:
+        if entity.relationships is not None:
+            entity_relationship_models.extend(
+                make_relationship_models(
+                    (entity.type_, entity.id_), entity.relationships, uploaded_entities
+                )
+            )
+
+    EntityDataModel.bulk_create(entity_relationship_models, 200)
```

Comment:
I think that what's happening here is that we are adding the entitydata.entity_id field more or less and using these relations to create the necessary joins etc. I'm starting to think that we're missing a layer between the pipeline functional code (i.e., take this input data/files and produce that output data/files) and the model layer (the simple data classes). The in-between layer could look roughly like an ORM but persisted in memory until ready to ship to the database.

Something to think about longer-term... IMO, a "command" should really just be glue code that glues together bits of the underlying classes / functions. What is here now feels like a bit more than that :D Especially in terms of connecting relationships / fetching data for exiting items.

Reply:
>The in-between layer could look roughly like an ORM but persisted in memory until ready to ship to the database.

This is a great idea. I'm fully behind it. As you noted, one of the clunky parts of the code as it stands is that many of these entities need to reference each other, though there will be thousands of rows to upload so it doesn't make sense to upload them all individually and fetch their row IDs whenever a foreign key relation needs to be established. I like this idea of a layer like the one you mentioned that takes care of establishing those relationships and performing the bulk upload once a paper is finished processing.

Reply:
>Something to think about longer-term... IMO, a "command" should really just be glue code that glues together bits of the underlying classes / functions.

100% agreed. The current command structure has ballooned in complexity beyond my initial expectations 🎈 😬 . I'd love to discuss further what the ideal structure looks like, and think about whether there's small ways we can remove complexity in future PRs as we make other changes before we're ready for a large-scale refactor.

Reply:
I think related to a previous conversation with Kyle, one thing that could be a step forward (possibly a bit large) would be some sort of entity DAO that does the following. Suppose I am running command Z that needs the output of command X and Y. Due to the pipeline wiring I've run X locally by Y only exists in the DB. The entity DAO makes this transparent to the command.

1. Check local disk for entities if there then yield to Z
2. Check DB for entities if there then yield to Z and probably also keep locally for other commands
3. Else, fail... Some sort of wiring or ordering is not quite right
4. OR instead of failing, invoke the necessary command (this seems like beyond the scope of the DAO)

You've modeled commands as a DAG, and I think that was the right call. Now, it's a question of making each command live in isolation.